### ISSUE TITLE
X-Forwarded-For input not being sanitized

### ISSUE THREAD

**Issue by ghost** [cherrypy/cherrypy#1368]
Originally reported by: **Anonymous**

---

If cherrypy is running in proxy-mode ('tools.proxy.on': True) the remote ip is blindly set to the first "ip" in the array that is returned by split() (see line 213 in cptools.py). This enables a malicious X-Forwarded-For Header to be used to set any value (even strings that aren't IPs) as cherrypy.request.remote.ip which might be a security vulnerabilty, depending on what the web application does with that information.

A POC is attached. If you access the page in your browser it should say 127.0.0.1; if you use for example curl to set the X-Forwarded-For Header (curl --header "X-Forwarded-For: test123" http://127.0.0.1:8080) you will get test123 as output.

---
- Bitbucket: https://bitbucket.org/cherrypy/cherrypy/issue/1368

**Comment by jaraco** [comment]
@harrim4n I agree with Eric that this isn't a critical issue, though if the security can be aligned to match best practices as seen in other web servers, such as the technique you reference with Apache, then that seems worthwhile, especially if it's unlikely to have negative consequences.

Would you be willing to draft an implementation?

**Comment by webknjaz** [comment]
@harrim4n ping

**Comment by seils** [comment]
Is there still an appetite to address this (i.e. validate the header is a valid IP address)?

**Comment by jaraco** [comment]
@seils Definitely. If you could draft an implementation similar to the one @harrim4n proposes, that would surely be acceptable.
